File: solutions/0054.go
package solutions

import (
	"fmt"
	"reflect"
)

const (
	Right = iota
	Down
	Left
	Up
)

func move(i, j int, dir *int, r, d, l, u *int) (int, int) {
	reti, retj := 0, 0

	switch *dir {
	case Right:
		nextj := j + 1
		if nextj == *r {
			*r = j
			*dir = Down
			reti = i + 1
			retj = j
		} else {
			reti = i
			retj = j + 1
		}
	case Down:
		nexti := i + 1
		if nexti == *d {
			*d = i
			*dir = Left
			reti = i
			retj = j - 1
		} else {
			reti = i + 1
			retj = j
		}
	case Left:
		nextj := j - 1
		if nextj == *l {
			*l = j
			*dir = Up
			reti = i - 1
			retj = j
		} else {
			reti = i
			retj = j - 1
		}
	case Up:
		nexti := i - 1
		if nexti == *u {
			*u = i
			*dir = Right
			reti = i
			retj = j + 1
		} else {
			reti = i - 1
			retj = j
		}
	}

	if (reti == *u || reti == *d) && (retj == *r || retj == *l) {
		reti = -1
		retj = -1
	}

	return reti, retj
}

func spiralOrder(matrix [][]int) []int {
	sizeX := len(matrix)
	if sizeX == 0 {
		return []int{}
	}
	sizeY := len(matrix[0])
	if sizeY == 0 {
		return []int{}
	}
	// dir
	r := sizeY
	d := sizeX
	l := -1
	u := 0
	i, j := 0, 0
	dir := Right
	ret := make([]int, 1)
	ret[0] = matrix[0][0]
	for {
		i, j = move(i, j, &dir, &r, &d, &l, &u)
		if i >= 0 && i < sizeX && j >= 0 && j < sizeY {
			fmt.Println("走一步: ", i, j, " 值为: ", matrix[i][j], " 方向: ", dir)
		} else {
			fmt.Println("异常i j: ", i, j, " 方向: ", dir)
		}
		if i == -1 && j == -1 {
			break
		} else {
			ret = append(ret, matrix[i][j])
		}
	}
	return ret
}

func init() {
	desc := `
Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.

Example 1:

Input:
[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
Output: [1,2,3,6,9,8,7,4,5]
Example 2:

Input:
[
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9,10,11,12]
]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
	`
	sol := Solution{
		Title:  "Spiral Matrix",
		Desc:   desc,
		Method: reflect.ValueOf(spiralOrder),
		Tests:  make([]TestCase, 0),
	}
	a := TestCase{}
	/*
		a.Input = []interface{}{[][]int{[]int{1, 2, 3}, []int{4, 5, 6}, []int{7, 8, 9}}}
		a.Output = []interface{}{[]int{1, 2, 3, 6, 9, 8, 7, 4, 5}}
		sol.Tests = append(sol.Tests, a)
	*/

	/*
		a.Input = []interface{}{[][]int{[]int{1, 2, 3}}}
		a.Output = []interface{}{[]int{1, 2, 3}}
		sol.Tests = append(sol.Tests, a)
	*/

	a.Input = []interface{}{[][]int{[]int{1}, []int{2}, []int{3}}}
	a.Output = []interface{}{[]int{1, 2, 3}}
	sol.Tests = append(sol.Tests, a)

	SolutionMap["0054"] = sol
}

File: solutions/0054.rs
/*
 * 给定一个包含 m x n 个元素的矩阵（m 行, n 列），请按照顺时针螺旋顺序，返回矩阵中的所有元素。
 *
 * 示例 1:
 *
 * 输入:
 * [
 *  [ 1, 2, 3 ],
 *  [ 4, 5, 6 ],
 *  [ 7, 8, 9 ]
 * ]
 * 输出: [1,2,3,6,9,8,7,4,5]
 * 示例 2:
 *
 * 输入:
 * [
 *   [1, 2, 3, 4],
 *   [5, 6, 7, 8],
 *   [9,10,11,12]
 * ]
 * 输出: [1,2,3,4,8,12,11,10,9,5,6,7]
 */

struct Solution {}

impl Solution {
    fn traverse_circle(
        matrix: &Vec<Vec<i32>>,
        i: usize,
        j: usize,
        m: usize,
        n: usize,
        output: &mut Vec<i32>,
    ) {
        if m == 1 {
            for v in 0..n {
                output.push(matrix[i][j + v]);
            }
            return;
        } else if n == 1 {
            for v in (0..m) {
                output.push(matrix[i + v][j]);
            }
            return;
        }

        let path_len = (m + n) * 2 - 4;
        let dest_one = n;
        let dest_two = m + n - 1;
        let dest_thr = m + n * 2 - 2;

        for d in 0..path_len {
            let val: i32;
            if d < dest_one {
                val = matrix[i][j + d]
            } else if d >= dest_one && d < dest_two {
                val = matrix[i + d - dest_one + 1][j + n - 1];
            } else if d >= dest_two && d < dest_thr {
                val = matrix[i + m - 1][j + n - 1 - (d - dest_two + 1)];
            } else if d >= dest_thr && d < path_len {
                val = matrix[i + m - 1 - (d - dest_thr + 1)][j];
            } else {
                continue;
            }

            output.push(val);
        }

        if m > 2 && n > 2 {
            Solution::traverse_circle(matrix, i + 1, j + 1, m - 2, n - 2, output);
        }
    }

    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let m = matrix.len();
        if m == 0 {
            return Vec::new();
        }
        let n = matrix[0].len();

        let mut spiral_order_vec = Vec::with_capacity(m * n);

        Solution::traverse_circle(&matrix, 0, 0, m, n, &mut spiral_order_vec);

        spiral_order_vec
    }
}

fn main() {
    println!("0054: ");

    //let input = vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]];
    let input = vec![vec![1, 2], vec![4, 5], vec![7, 8]];
    //let input = vec![vec![1], vec![4], vec![7]];
    //let input = vec![vec![1, 2], vec![3, 4]];
    let output = Solution::spiral_order(input);
    println!("spriral-order: {:?}", output);
}
